test/Network/IpAddress.cpp

```diff
@@ -0,0 +1,223 @@
+#include <SFML/Network/IpAddress.hpp>
+
+#include <doctest.h>
+
+#include <sstream>
+
+using namespace std::string_literals;
+
+TEST_CASE("sf::IpAddress class - [network]")
+{
+    SUBCASE("Construction")
+    {
+        SUBCASE("Default constructor")
+        {
+            const sf::IpAddress ipAddress;
+            CHECK(ipAddress.toString() == "0.0.0.0"s);
+            CHECK(ipAddress.toInteger() == 0);
+            CHECK(ipAddress != sf::IpAddress::Any);
+        }
+
+        SUBCASE("std::string constructor")
+        {
+            const sf::IpAddress ipAddress = "192.168.0.1"s;
+            CHECK(ipAddress.toString() == "192.168.0.1"s);
+            CHECK(ipAddress.toInteger() == 0xC0A80001);
+            CHECK(ipAddress != sf::IpAddress::None);
+
+            const sf::IpAddress broadcast("255.255.255.255"s);
+            CHECK(broadcast.toString() == "255.255.255.255"s);
+            CHECK(broadcast.toInteger() == 0xFFFFFFFF);
+            CHECK(broadcast == sf::IpAddress::Broadcast);
+
+            const sf::IpAddress any("0.0.0.0"s);
+            CHECK(any.toString() == "0.0.0.0"s);
+            CHECK(any.toInteger() == 0x00000000);
+            CHECK(any == sf::IpAddress::Any);
+
+            const sf::IpAddress localHost("localhost"s);
+            CHECK(localHost.toString() == "127.0.0.1"s);
+            CHECK(localHost.toInteger() == 0x7F000001);
+            CHECK(localHost == sf::IpAddress::LocalHost);
+
+            const sf::IpAddress invalidIpAddress("255.255.255.256"s);
+            CHECK(invalidIpAddress.toString() == "0.0.0.0"s);
+            CHECK(invalidIpAddress.toInteger() == 0);
+            CHECK(invalidIpAddress == sf::IpAddress::None);
```

Comment:
Side note: this might also a bit unexpected, but I don't really see how we can fix it, as long as error handling is not sorted out. The SFML 2 resource approach of `sf::Image im; if (!im.loadFromY(...))` is too clumsy for such a simple type. There are a few options (`optional`, exceptions) but we should probably postpone that until we find a general way to address error handling.

Reply:
This is a textbook use case for exceptions since signaling constructor failure is very hard through other means. I'm of the opinion that exceptions are a good fit for SFML but I know the team made the very specific reason not to use them so I see no better way to signal failure. Maybe in these cases we also write to `sf::err()` to signal an error occurred? That would be in line with other "might fail" APIs.

Reply:
What @vittorioromeo mentioned is btw a very common idiom in Rust. There are no constructors in the language (and no exceptions), just "static factory" methods, which return either the object itself or a wrapper object for fallible construction:
* `Option<T>` which is pretty much 1:1 `std::optional`
* `Result<T, E>` which contains either a value or an error (there's a [proposal for `std::expected`](https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4015.pdf)).

Of course, the situation is different because the language was designed with those in mind and it's thus "idiomatic" by definition, while in C++, fallible factory methods are just one of many approaches. I don't mean this as a way to say Rust is right here, more that there's prior art for this idiom, which has worked well on a large scale.

I think this topic needs more discussion outside this PR; we can also leave the error handling as-is for now, until we discussed the topic more thoroughly.

Reply:
> We cannot easily signal why the conversion failed

A good short term solution for a later PR is to add some `sf::err` logging. We can add unit tests for this as well. I implemented this locally and it works just fine.